File: 15. Merge k sorted arrays using pq.cpp
/*
 Task: Merge k sorted arrays using a priority queue
*/

#include <bits/stdc++.h>
using namespace std;

int main()
{
    vector<vector<int>> lists = {{1, 4, 7}, {2, 5, 8}, {3, 6, 9}};
    priority_queue<pair<int, pair<int, int>>, vector<pair<int, pair<int, int>>>, greater<>> pq;

    for (int i = 0; i < static_cast<int>(lists.size()); ++i)
    {
        if (!lists[i].empty())
            pq.push({lists[i][0], {i, 0}});
    }

    vector<int> out;
    while (!pq.empty())
    {
        auto t = pq.top();
        pq.pop();
        out.push_back(t.first);
        int i = t.second.first;
        int idx = t.second.second;
        if (idx + 1 < static_cast<int>(lists[i].size()))
            pq.push({lists[i][idx + 1], {i, idx + 1}});
    }

    for (int x : out) cout << x << " ";
    cout << "\n";
    return 0;
}

// Time Complexity: O(N log k)
// Space Complexity: O(k)

File: 15. Merge k sorted arrays using pq.py
"""
Task: Merge K sorted arrays using a min-heap
Time Complexity: O(n log k)
Space Complexity: O(k)
"""

import heapq


def merge_k_arrays(lists):
    pq = []
    for i, arr in enumerate(lists):
        if arr:
            heapq.heappush(pq, (arr[0], i, 0))
    res = []
    while pq:
        val, i, j = heapq.heappop(pq)
        res.append(val)
        if j+1 < len(lists[i]):
            heapq.heappush(pq, (lists[i][j+1], i, j+1))
    return res

if __name__ == "__main__":
    print(merge_k_arrays([[1,4,5],[1,3,4],[2,6]]))

# Task: Merge k sorted arrays using pq
